jupyterhub/metrics.py

```diff
@@ -0,0 +1,41 @@
+"""
+Prometheus metrics exported by JupyterHub
+"""
+from prometheus_client import Histogram
+
+REQUEST_DURATION_SECONDS = Histogram(
+    'request_duration_seconds',
+    'request duration for all HTTP requests',
+    ['method', 'handler', 'code']
+)
+
+SPAWN_DURATION_SECONDS = Histogram(
+    'spawn_duration_seconds',
+    'spawn duration for all server spawns',
+    ['status'],
+    buckets=[0.5, 1, 2.5, 5, 10, 15, 30, 60, 120, float("inf")]
+)
+
+PROXY_ADD_DURATION_SECONDS = Histogram(
+    'proxy_add_duration_seconds',
+    'duration for adding user routes to proxy',
+    ['status']
+)
+
+def prometheus_log_method(handler):
+    """
+    Tornado log handler for recording RED metrics
```

Comment:
Is this more accurate?  Log RED metrics as a bit more information to Tornado's default log

Reply:
It doesn't add to the tornado default log, rather it is just a convenient hook function that lets us read timing & request info. I've added a couple of lines here explaining how this is used!